Update PR when labeled/opened

f1sherman opened the pull request:
This will conditionally update a PR whenever the PR is labeled or opened, which will partially fulfill [this
suggestion](https://github.com/palantir/bulldozer/issues/145#issuecomment-520662592) for https://github.com/palantir/bulldozer/issues/145.

It doesn't seem possible to determine which label was added via the Github webhook payload, so this will update the PR regardless of whether the label added is what is configured in `bulldozer.yml`.

I haven't had a chance to test this yet (will update when I have), so I'm just looking for early feedback at this point. Thanks!

f1sherman committed:
Update PR when labeled/opened

This will conditionally update a PR whenever the PR is labeled or opened, which
will partially fulfill [this
suggestion](https://github.com/palantir/bulldozer/issues/145#issuecomment-520662592)
for https://github.com/palantir/bulldozer/issues/145.

It doesn't seem possible to determine which label was added via the
Github webhook payload, so this will update the PR regardless of whether
the label added is what is configured in `bulldozer.yml`.

bluekeyes commented on server/handler/pull_request.go:
```diff
@@ -66,6 +67,21 @@ func (h *PullRequest) Handle(ctx context.Context, eventType, deliveryID string,
 	}
 
+	if event.GetAction() == "labeled" || event.GetAction() == "opened" {
+		logger.Debug().Msgf("received pull_request %s event", event.GetAction())
```
This logging is probably more appropriate higher up in the function where it will apply for all possible actions, rather than only `labeled` and `opened` actions.

f1sherman replied:
Great idea, will move this further up

bluekeyes commented on server/handler/pull_request.go:
```diff
@@ -66,6 +67,21 @@ func (h *PullRequest) Handle(ctx context.Context, eventType, deliveryID string,
+		}
+
+		baseRef := ref.GetObject().GetSHA()
+		if err := h.UpdatePullRequest(logger.WithContext(ctx), pullCtx, client, pr, baseRef); err != nil {
```
The `baseRef` passed to `UpdatePullRequest` is a ref name rather than a SHA. You should be able to use the `fmt.Sprintf("refs/heads/%s", base)` directly, without resolving it.

f1sherman replied:
Nice, that will really simplify things

bluekeyes reviewed (commented):
Thanks for adding this. Based on the Go documentation for [`PullRequestEvent`](https://pkg.go.dev/github.com/google/go-github/v32/github#PullRequestEvent), there is a `Label` field that is populated for `labeled` actions. That said, the current structure of the code makes it difficult to use this value for anything, so I think we can try the simple approach of doing this evaluation on every `labeled` action for now.

The bigger refactor (which you might not want to take on) would be to change `Base#ProcessPR` and `Base#UpdatePR` so that we don't duplicate configuration loading. As is, we'll look for and load configuration twice, although the requests the second time should hit the HTTP cache.

f1sherman committed:
Only fetch config once during `pull_request` event

Refactor config fetching to happen before calling `ProcessPR` and
`UpdatePR` so that it isn't fetched twice during the `pull_request`
event when we're both merging and updating a PR.

f1sherman committed:
Update PR before trying to merge it

It probably makes sense to do things in this order rather than how we
were doing it before.

f1sherman commented:
@bluekeyes I believe I've addressed all of your feedback. I also took a shot at that refactor as a separate commit, let me know if you want me to change that or remove it. Thanks!

bluekeyes commented on server/handler/base.go:
```diff
@@ -33,14 +33,28 @@ type Base struct {
 
+	switch {
+	case bulldozerConfig.Missing():
+		return nil, errors.Errorf("No configuration found for %s", bulldozerConfig)
```
One consequence of this conversion is that these conditions that were previously only logged are now returned and propagated as errors. I think I'd like to keep the old behavior by logging the same messages here and then returning `nil, nil`.

This means that either callers will have to check for `nil` or `UpdatePR` and `ProcessPR` should return immediately if passed a `nil` configuration; I'd probably have the functions return immediately.

bluekeyes commented on server/handler/base.go:
```diff
@@ -33,14 +33,28 @@ type Base struct {
+	}
+
+	logger.Debug().Msgf("Found valid configuration for %s", bulldozerConfig)
+	config := bulldozerConfig.Config
+
+	return config, nil
```
style:
```suggestion
	return bulldozerConfig.Config, nil
```

bluekeyes commented on server/handler/check_run.go:
```diff
@@ -73,7 +73,11 @@ func (h *CheckRun) Handle(ctx context.Context, eventType, deliveryID string, pay
-		if err := h.ProcessPullRequest(logger.WithContext(ctx), pullCtx, client, fullPR); err != nil {
+		config, err := h.FetchConfig(ctx, client, pr)
+		if err != nil {
+			return errors.Wrap(err, "failed to fetch configuration")
```
Because `FetchConfig` already wraps errors with a similar message, wrapping them again doesn't add useful context.
```suggestion
			return err
```

bluekeyes commented on server/handler/issue_comment.go:
```diff
@@ -57,7 +57,11 @@ func (h *IssueComment) Handle(ctx context.Context, eventType, deliveryID string,
-	if err := h.ProcessPullRequest(ctx, pullCtx, client, pr); err != nil {
+	config, err := h.FetchConfig(ctx, client, pr)
+	if err != nil {
+		return errors.Wrap(err, "failed to fetch configuration")
```
```suggestion
		return err
```

bluekeyes commented on server/handler/pull_request.go:
```diff
@@ -62,7 +64,19 @@ func (h *PullRequest) Handle(ctx context.Context, eventType, deliveryID string,
-	if err := h.ProcessPullRequest(ctx, pullCtx, client, pr); err != nil {
+	config, err := h.FetchConfig(ctx, client, pr)
+	if err != nil {
+		return errors.Wrap(err, "failed to fetch configuration")
```
```suggestion
		return err
```

bluekeyes commented on server/handler/status.go:
```diff
@@ -68,7 +68,11 @@ func (h *Status) Handle(ctx context.Context, eventType, deliveryID string, paylo
-		if err := h.ProcessPullRequest(logger.WithContext(ctx), pullCtx, client, pr); err != nil {
+		config, err := h.FetchConfig(ctx, client, pr)
+		if err != nil {
+			return errors.Wrap(err, "failed to fetch configuration")
```
```suggestion
			return err
```

bluekeyes reviewed (commented):
Thanks for taking on the config refactor too. All of my comments relate to that; the PR handler part of this change looks good. If you'd like to fix up the refactor part, that would be great, otherwise, feel free to remove that commit and I'll clean up some of the duplication in a separate change.

f1sherman commented:
@bluekeyes ok I went ahead and made those updates. Thanks so much for patiently working with me through those. Is this something you'd like me to test before merging or would it make more sense to merge first then test? Merging first makes testing a little easier for me but I'm ok with either way.

bluekeyes commented:
This seems relatively safe in terms of changes to existing functionality, so I think we can merge first and then test.

f1sherman commented:
@bluekeyes just an update that I've been testing this today and it's working well!

f1sherman commented:
@bluekeyes after some more testing I noticed a bug with this change (or at least I consider it a bug): when a webhook triggers both an update and a merge, Bulldozer continues to use the stale statuses from before the update to evaluate whether a merge can happen. I'll work on putting together a PR to fix this.

f1sherman commented:
@bluekeyes I put together a proposed fix for this issue [here](https://github.com/palantir/bulldozer/pull/242). Feedback welcome (as always)!
